### PR TITLE
Added tie, aka. assignalbe tuples

### PR DESCRIPTION
**Pull request by Bolpat**
With `tie` you can deconstruct tuples in a convenient manner, that serves when functions
return tuples, you want to swap values or simultaneously iterate variables.

Note, however, that swapping with `tie` is basically using intermediate variables,
in contrast to $(REF swap, std, algorithm, mutation).

You can use `$` or `null` on the left-hand side to ignore right-hand side
values. Use `tie` on the right-hand side for heterogeneous tuples; for
type-homogeneous tuples, static arrays are fine, too.

```d
int x = 2, y = 3;

// Swapping.
tie[ x, y ] = tie[ y, x ];
assert(x == 3 && y == 2);
// Simultaneous iteration.
tie[ x, y ] = tie[ y + 1, x * 2 ];
assert(x == 3 && y == 6);
// If all types support an operator op, you can use op=, too.
tie[ x, y ] += tie[ 2, 1 ];
assert(x == 5 && y == 7);

// $ is used for discarding.
tie[ $, x, $, y, $ ] = tie[ 0, 1, 2, 3, 4 ];
assert(x == 1 && y == 3);

static auto returnsTuple(int v) { return tuple(v * 2, v + 2); }
static int[3] returnsStaticArray(int v) { return [ v - 1, v, v + 1 ]; }

// Deconstruction.
tie[ x, y ] = returnsTuple(3);
assert(x == 6 && y == 5);
tie[ x, $, y ] = returnsStaticArray(x);
assert(x == 5 && y == 7);
```

`tie` supports non-copyable and must-initialize types.

### CODE REVIEW

**Comment by Bolpat** [comment]
I don't know why CircleCI fails. It says
```
std/typecons.d(2662:9)[warn]: Left side of logical or is identical to right side.
```
My commit does not touch that line. In the mentioned line, there is no logical or either.

**Comment by n8sh** [comment]
For the benefit of people who aren't coming from the C++ world, this PR seems to be mirroring [`std::tie` which was added in C++11](http://en.cppreference.com/w/cpp/utility/tuple/tie).

**Comment by Bolpat** [comment]
One thing `tie` cannot do, is declaring variables. There is a work-in-progress first-class tuple DIP that suggests introducing `auto (a, b) = returnsTuple()` and `(auto a, auto b) = returnsTuple()` to declare when unpacking.

This could be done with a much more general language change: `out` declarations. If a function `void f(out int x)` has a `out` parameter, one can then use
```d
f(out int x);
```
to declare a variable `x`. For templates, we'd need `auto out`.

With that, `tie` can fully replace first-class tuples.

In current D, as any value can be moved, you can always declare a variable beforehand and use `tie` to move something into it. With DIP 1014, this could be false in general.

**Comment by Bolpat** [comment]
I have some questions to you who are reading this:
1. Should `tie` support array literals on the right-hand side?
2. Should `tie` support rubber-discard? You'd use `tie[x, $..$, y] = returns4Tuple()` instead of `tie[x, $, $, y] = ...`. Obviously, you can use only one rubber-discard.
3. Should `tie` support tuple unpacking on the last position on the left-hand side? Say, you have `int x` and `Tuple!(int, int) t` use `tie[x, t] = returns3Tuple()`.
4. Should `tie` support tuple unpacking on any position on the left-hand side? Say, you have `int x, y` and `Tuple!(int, int) t` use `tie[x, t, y] = returns4Tuple()`.

Comments about
1. It would be easy to implement assigning from array literals, i.e. `tie[ ... ] = [ ... ]`; the problem being implicit conversion inside the right-hand side literal might be overseen.
2. This abuses `$` even more. In that context, it means something completely different than with ranges. I think, `$` for discard is acceptable, it's hard to misinterpret. The following are even worse.
3. Could be difficult to understand. I tried an approach enforcing `tie[x, t..$] = ...`, but that necessitates pointers. It'd be easy to decay `t..$` into `t`, making `..$` optional.
4. See 3.

**Comment by dukc** [comment]
This sure is a big decision, if we are to allow this in principle. Except a bloody debate. A lot of great use cases, definitely.

ping @andralex

File: std/typecons.d [hunk 1]
```diff
@@ -2094,6 +2100,343 @@ enum isTuple(T) = __traits(compiles,
...
+        return s[0 .. $ - 1];
+    }
+    enum string hw = ctfeFunction(); // evaluated at compile-time
+    static assert(hw == "Hello, World!");
+}
+
+/// `tie` supports moving. It can be observed using an explicit destructor.
+pure nothrow @nogc @safe
+unittest
+{
```
Comment (dukc) [inline std/typecons.d]: Thats one long unittest. Can it be split?

File: std/typecons.d [hunk 2]
```diff
@@ -2094,6 +2100,343 @@ enum isTuple(T) = __traits(compiles,
...
+    assert(x == 2);
+    assert(mtuple[0].isInit); // ... from here
+    assert(mtuple[1].isInit); // destroyed
+    // Tuples do not support piecewise move extraction
+}
+
+/// `tie` supports non copyable and must-initialize types.
+pure nothrow @nogc @safe
+unittest
+{
```
Comment (dukc) [inline std/typecons.d]: Again, a long unittest.

File: std/typecons.d [hunk 3]
```diff
@@ -2094,6 +2100,343 @@ enum isTuple(T) = __traits(compiles,
...
+    tie[ nc ] = cast(NoCopy[1]) [ move(nc2) ];
+    assert(nc == 3 && nc2.isInit);
+    tie[ nc, nc2 ] = f();
+    assert(nc == 1 && nc2 == 2);
+
+    NoCopy[2] ncarray = [ NoCopy(1), NoCopy(2) ];
+    static assert(!__traits(compiles, {
+        tie[ nc, $ ] = ncarray; // this copies!
+    }));
+    // These are supposed to work; see Limitations.
```
Comment (dukc) [inline std/typecons.d]: Hmm, if we have nothing against them working, why unittest that they won't?

Comment (Bolpat) [reply std/typecons.d]: As soon as the bug is fixed, they won't compile and the documentation can be adapted.
If that's not intended, I'll remove it.

**Comment by Bolpat** [comment]
I cannot merge because on the active master
```
dmd -main -unittest -run std/typecons.d
```
fails with linker error:
```
Error 42: Symbol Undefined __D3std4conv16testEmplaceChunkFNaNbNiNfAvkkZv
```

**Comment by Bolpat** [comment]
Sorry, wrong button click.

File: std/typecons.d (hunk 1 above)
Comment (Bolpat) [reply std/typecons.d]: I have split it.

File: std/typecons.d (hunk 2 above)
Comment (Bolpat) [reply std/typecons.d]: Yes, in contrast to the other, is one is concerned about must-initialize and non-copyable types. Splitting it means duplicating the struct. The unit test is long, splitting it makes it longer. There's nothing to be left out. If you're dealing with non-copyable types, there's enough struggle. At least, the examples here will be rich.

File: std/typecons.d (hunk 3 above)
Comment (dukc) [reply std/typecons.d]: Good rationale.

**Comment by Bolpat** [comment]
Closing, no answers in years.

**Pull request closed without merging**